ncclient/operations/rpc.py

```diff
@@ -208,23 +208,26 @@ def callback(self, root, raw):
         if self._device_handler.perform_qualify_check():
             if tag != qualify("rpc-reply"):
                 return
-        for key in attrs: # in the <rpc-reply> attributes
-            if key == "message-id": # if we found msgid attr
-                id = attrs[key] # get the msgid
-                with self._lock:
-                    try:
-                        rpc = self._id2rpc[id] # the corresponding rpc
-                        logger.debug("Delivering to %r" % rpc)
-                        rpc.deliver_reply(raw)
-                    except KeyError:
-                        raise OperationError("Unknown 'message-id': %s" % id)
-                    # no catching other exceptions, fail loudly if must
-                    else:
-                        # if no error delivering, can del the reference to the RPC
-                        del self._id2rpc[id]
-                        break
-        else:
-            raise OperationError("Could not find 'message-id' attribute in <rpc-reply>")
+        for key in attrs:  # in the <rpc-reply> attributes
+            message_id = attrs.get("message-id")
+            if message_id is None:
+                # required attribute so raise OperationError
+                raise OperationError("Could not find 'message-id' attribute in <rpc-reply>")
```

Comment:
This check should be done before entering the `for` loop, I think? Doesn't make much sense to repeat it for every key.

Reply:
For that matter, I'm not sure why there's a for loop here at all - it looks like message-id is the only key we care about?

Reply:
Sorry - also would `if "message-id" not in attrs` be a simpler way to perform this check?

Reply:
Incorporated your comments to move the check out of the for loop and do it in a simpler way. Waiting for Einar's thoughts on removing the for loop altogether.